### PR TITLE
[MKL][BLAS] Add omatcopy2 interface

### PR DESCRIPTION
**Pull request by s-Nick**
# Description

This PR add interface and test for `omatcopy2` operator. The operator is part of the oneAPI spec but not implemented yet. We implemented in portBLAS backend and so we are adding the interface for it now. 

# Checklist

## All Submissions

- [x] Do all unit tests pass locally?
  Tested multiple backends:
 
[cublas_all.txt](https://github.com/oneapi-src/oneMKL/files/13174886/cublas_all.txt)
[mklgpu_all.txt](https://github.com/oneapi-src/oneMKL/files/13174905/mklgpu_all.txt)
[mklcpu_all.txt](https://github.com/oneapi-src/oneMKL/files/13174909/mklcpu_all.txt)
[portBLAS_all.txt](https://github.com/oneapi-src/oneMKL/files/13174913/portBLAS_all.txt)
[rocblas_all.txt](https://github.com/oneapi-src/oneMKL/files/13174918/rocblas_all.txt)

- [x] Have you formatted the code using clang-format?

## New interfaces

- [x] Have you provided motivation for adding a new feature as part of RFC and
it was accepted? # (RFC)
  This interface was already added and accepted in [#420](https://github.com/oneapi-src/oneAPI-spec/pull/420)
- [x] What version of oneAPI spec the interface is targeted?
  (1.2 provisional rev 1)

## New features

- [x] Have you provided motivation for adding a new feature?
- [x] Have you added relevant tests?
  Added tests are skipped for backends that do not offer this interface. To test its correctness added also a cpu reference implementation as it is for `omatcopy`, `imatcopy` and `omatadd`.

Test with multiple backends:
 
[portBLAS_omatcopy2.txt](https://github.com/oneapi-src/oneMKL/files/13175037/portBLAS_omatcopy2.txt)

[cublas_omatcopy2.txt](https://github.com/oneapi-src/oneMKL/files/13175026/cublas_omatcopy2.txt)
[mklgpu_omatcopy2.txt](https://github.com/oneapi-src/oneMKL/files/13175031/mklgpu_omatcopy2.txt)
[mklcpu_omatcopy2.txt](https://github.com/oneapi-src/oneMKL/files/13175030/mklcpu_omatcopy2.txt)
[rocblas_omatcopy2.txt](https://github.com/oneapi-src/oneMKL/files/13175042/rocblas_omatcopy2.txt)

### CODE REVIEW

File: tests/unit_tests/blas/include/reference_blas_templates.hpp [hunk 1]
```diff
@@ -2011,6 +2013,43 @@ void omatcopy_ref(oneapi::mkl::layout layout, oneapi::mkl::transpose trans, int6
...
+        logical_m = m;
+        logical_n = n;
+    }
+    else {
+        logical_m = n;
+        logical_n = m;
+    }
+    if (trans == oneapi::mkl::transpose::trans) {
+        for (int64_t i = 0; i < logical_m; ++i) {
+            for (int64_t j = 0, c = 0; j < logical_n; ++j, ++c) {
```
Comment (OuadiElfarouki) [inline tests/unit_tests/blas/include/reference_blas_templates.hpp]: Is the `c` of any use here (and in 2042) ?

**Review by muhammad-tanvir-1211 — APPROVED** [review]

File: tests/unit_tests/blas/include/reference_blas_templates.hpp (hunk 1 above)
Comment (s-Nick) [reply tests/unit_tests/blas/include/reference_blas_templates.hpp]: Thank you for your feedback @OuadiElfarouki , I addressed it in cfcfab70136a06862c43b7d4d9e7d0b762262139

File: tests/unit_tests/blas/extensions/omatcopy2.cpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
+/*******************************************************************************
+* Copyright 2022 Intel Corporation
```
Comment (andrewtbarker) [inline tests/unit_tests/blas/extensions/omatcopy2.cpp]:
```suggestion
* Copyright 2023 Intel Corporation
```

File: tests/unit_tests/blas/extensions/omatcopy2_usm.cpp [hunk 3]
```diff
@@ -0,0 +1,205 @@
+/*******************************************************************************
+* Copyright 2022 Intel Corporation
```
Comment (andrewtbarker) [inline tests/unit_tests/blas/extensions/omatcopy2_usm.cpp]:
```suggestion
* Copyright 2023 Intel Corporation
```

**Review by andrewtbarker — APPROVED** [review]
After resolving conflicts and minor copyright fixes this looks good to me.

File: tests/unit_tests/blas/extensions/omatcopy2_usm.cpp (hunk 3 above)
Comment (s-Nick) [reply tests/unit_tests/blas/extensions/omatcopy2_usm.cpp]: Thank you for your review @andrewtbarker. I updated the copyrights and fixed the conflicts in the last commit.

**Review by mkrainiuk — APPROVED** [review]

**Pull request merged**